lib.path.append: init

infinisil opened the pull request:
###### Description of changes

This function can be used to append strings to Nix path values in a safe way, it's like `<path> + ("/" + <string>)` but safer. Relates to other work in [the path library effort](https://github.com/NixOS/nixpkgs/issues/210426).

This work is sponsored by [Antithesis](https://antithesis.com/) :sparkles:

###### Things done

- [x] Wrote documentation
- [x] Added and ran unit tests

roberth commented on lib/path/default.nix:
```diff
@@ -94,6 +95,47 @@ let
 
+  /* Append a subpath string to an absolute path value, the result is an absolute path value.
+
+    Like `<path> + ("/" + <string>)` but safer, because it errors for potentially unexpected results.
```
```suggestion
    Like `<<path>> + ("/" + <<string>>)` but safer, because it errors for potentially unexpected results.
```

`<path>` is syntax for `NIX_PATH`.

infinisil replied:
I changed it to just `path + ("/" + string)` now, I think that's nicer

roberth reviewed (commented):
Otherwise looks alright.

fricklerhandwerk commented on lib/path/default.nix:
```diff
@@ -94,6 +95,47 @@ let
 
+  /* Append a subpath string to an absolute path value, the result is an absolute path value.
+
+    Like `path + ("/" + string)` but safer, because it errors for potentially unexpected results.
```
```suggestion
    Like `path + ("/" + string)` but safer, because it errors instead of returning potentially surprising results.
```

fricklerhandwerk commented on lib/path/default.nix:
```diff
@@ -94,6 +95,47 @@ let
+
+    Like `path + ("/" + string)` but safer, because it errors for potentially unexpected results.
+    More specifically, it checks that the first argument is a path value,
+    and that the second argument is a valid subpath string, see `lib.path.subpath.isValid`.
```
Add a link to the manual section so it can be navigated to easily.

infinisil replied:
Unfortunately there's no good way to link to it, markdown syntax here doesn't get rendered. I guess I could dig into making that work

fricklerhandwerk commented on lib/path/default.nix:
```diff
@@ -94,6 +95,47 @@ let
+  /* Append a subpath string to an absolute path value, the result is an absolute path value.
+
+    Like `path + ("/" + string)` but safer, because it errors for potentially unexpected results.
+    More specifically, it checks that the first argument is a path value,
```
```suggestion
    More specifically, it checks that the first argument is a [path],
```

Link to the Nix manual to make clear what we mean.
Why do we say "path value"? Is there something specific about the value?

infinisil replied:
I called it "path value" because "path" is a very ambiguous term. With "value" it becomes much clearer that this is a Nix type. But I guess if we link to it that takes care of it too.

fricklerhandwerk commented on lib/path/default.nix:
```diff
@@ -94,6 +95,47 @@ let
+      append /. "foo/bar"
+      => /foo/bar
+
+      # first argument needs to be a Nix path value
```
```suggestion
      # first argument needs to be a Nix path
```

infinisil replied:
This might be especially confusing because the next line errors for `append "/foo" "bar"`, where the first argument does indeed look like a "path". Maybe it should be

```suggestion
      # first argument needs to be a path data type
```

instead

fricklerhandwerk commented on lib/path/default.nix:
```diff
@@ -94,6 +95,47 @@ let
+  */
+  append = base: subpath:
+    assert assertMsg (isPath base)
+      "lib.path.append: The first argument is of type ${builtins.typeOf base}, but a path value was expected";
```
```suggestion
      "lib.path.append: The first argument is of type ${builtins.typeOf base}, but a path was expected";
```

fricklerhandwerk commented on lib/path/default.nix:
```diff
@@ -94,6 +95,47 @@ let
+  append = base: subpath:
+    assert assertMsg (isPath base)
+      "lib.path.append: The first argument is of type ${builtins.typeOf base}, but a path value was expected";
+    assert assertMsg (subpathInvalidReason subpath == null)
+      "lib.path.append: Second argument is not a valid subpath string: ${subpathInvalidReason subpath}";
```
```suggestion
    assert assertMsg (subpathInvalidReason subpath == null) ''
      lib.path.append: Second argument is not a valid subpath string:
          ${subpathInvalidReason subpath}
    '';
```

infinisil replied:
Previously discussed in https://github.com/NixOS/nixpkgs/pull/205190#discussion_r1044936046. At least `path.normalise` and all the other library functions do not use newlines in throws, I think we should stay consistent with that.

roberth replied:
Exceptions are only ever presented one at a time, so the need for consistency isn't really there.
Nix itself uses multi-line messages, so feel free to do the same.

fricklerhandwerk replied:
In terms of consistency, we can of course fix up the code in multiple places in a follow-up. I'd just prefer to start while creating new code instead of piling up more work for later, which is more probable to get glanced over.

infinisil replied:
Fair enough, I'm using newlines now

fricklerhandwerk reviewed (approved):
Just a few comments on cosmetics. Good stuff, and really compact thanks to a solid basis.

infinisil commented:
Made some minor changes to the documentation. Using "path value type" when it could be ambiguous or being more precise is useful. And I'm not using any markdown syntax in the docs because it's not rendered at all. I looked into it a bit, but it's not very trivial to get it to render.

roberth commented:
> I'm not using any markdown syntax in the docs because it's not rendered at all.

I think it's fine to write markdown even if it's not rendered, but for now it really doesn't matter.

roberth commented on lib/path/tests/unit.nix:
```diff
@@ -3,9 +3,42 @@
+      expected = false;
+    };
+    testAppendExample5 = {
+      expr = (builtins.tryEval (append /foo /bar)).success;
```
Maybe reference the design document or add a short explanation?
Same for `"/bar"` below.

infinisil replied:
The documentation for `append` already has short explanations for all the examples, and the tests here just verify the output of those examples. I'll add a comment on the top to point out that these "Example[0-9]+" test cases correspond to the examples in the documentation.

infinisil commented on lib/path/default.nix:
```diff
@@ -94,6 +95,48 @@ let
+  */
+  append = base: subpath:
+    assert assertMsg (isPath base) ''
+      lib.path.append: The first argument is of type ${builtins.typeOf base}, but a path was expected'';
```
I'm putting `''` at the end of the string here and not on a newline because we don't want to print an extra newline (`throw` adds one). Apparently it doesn't matter though, because `throw` removes all trailing newlines anyways:

```
nix-repl> throw "hello\n\n\n\n\n"
error: hello

nix-repl>
```

I don't want to rely on that because it's undocumented and perhaps a bug. While I'm at it, there's a bug where `throw ""` doesn't print the red `error:` prefix:

```
nix-repl> throw ""                


nix-repl>
```

infinisil commented:
> > I'm not using any markdown syntax in the docs because it's not rendered at all.
> 
> I think it's fine to write markdown even if it's not rendered, but for now it really doesn't matter.

I guess a good policy for now is: "Use markdown syntax when it doesn't make it significantly harder to read without being rendered". I'm using that here now

infinisil committed:
lib.path: Minor improvements

- Use isValid when possible instead of subpathInvalidReason: https://github.com/NixOS/nixpkgs/pull/209099#discussion_r1068714681
- Add documentation to function arguments
- Use newlines for error messages: https://github.com/NixOS/nixpkgs/pull/208887#discussion_r1069737602
- Add short comments for the unit test groups: https://github.com/NixOS/nixpkgs/pull/208887#discussion_r1072913051
- Slight formatting improvement for laws: https://github.com/NixOS/nixpkgs/pull/209099#discussion_r1068707955

infinisil committed:
lib.path.append: init

This function can be used to append strings to Nix path values in a
safe way.

infinisil commented:
Now consistently using `isValid` and newlines in errors for all path library functions now, also added argument documentation.

infinisil commented:
Now even more consistency with the other PRs. I think this is ready to be merged

fricklerhandwerk reviewed (approved):
As always, amazing work. ✨ Thanks a lot for your patience with the reviews.

infinisil commented:
@roberth Does it also look good to you? If so I'd appreciate a mergin :)
